### PR TITLE
Migrate DetailsLink component

### PR DESCRIPTION
**Pull request by reykjalin**
Fixes part of #1351 

#### Changes proposed in this Pull Request

- Migrate `<DetailsLink />` and corresponding tests to TypeScript.
- Add declarations for the `<Link />` component from `@woocommerce/components`.
- Add declarations for the `<Gridicon />` component from `gridicons`.

#### Testing instructions

**Make sure to run `npm ci` and `npx webpack` (or `npm start`) before you start**

- Run `npm run test:js -- client/components/details-link` and make sure tests pass
- Open any of the tables that use the `<DetailsLink />` component (e.g. the deposits list) and make sure the `<DepositsLink />` renders correctly.

-------------------

- [x] Added changelog entry (or does not apply)
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (or does not apply)

### CODE REVIEW

File: .eslintrc [hunk 1]
```diff
@@ -96,7 +96,8 @@
 				"@typescript-eslint/no-explicit-any": "off",
 				"no-use-before-define": "off",
 				"@typescript-eslint/no-use-before-define": [ "error" ],
-				"react/react-in-jsx-scope": "error"
+				"react/react-in-jsx-scope": "error",
+				"@typescript-eslint/explicit-module-boundary-types": "off"
```
Comment (shendy-a8c) [inline .eslintrc:100]: Why setting this off?

File: client/components/details-link/index.tsx [hunk 2]
```diff
@@ -3,18 +3,24 @@
...
  * External dependencies
  */
-import Gridicon from 'gridicons';
+import * as React from 'react';
+import * as Gridicon from 'gridicons';
 import { addQueryArgs } from '@wordpress/url';
 import { Link } from '@woocommerce/components';
 
-export const getDetailsURL = ( id, parentSegment ) =>
+export const getDetailsURL = ( id: string, parentSegment: string ) =>
```
Comment (shendy-a8c) [inline client/components/details-link/index.tsx:11]: With rule `@typescript-eslint/explicit-module-boundary-types` on, linter complains this exported function is missing return type.

```suggestion
export const getDetailsURL = ( id: string, parentSegment: string ): string =>
```

File: client/components/details-link/index.tsx [hunk 3]
```diff
@@ -3,18 +3,24 @@
...
 		id,
 	} );
 
-const DetailsLink = ( { id, parentSegment } ) =>
+type DetailsLinkParams = {
+	id?: string;
+	parentSegment: string;
+};
+
+const DetailsLink = ( { id, parentSegment }: DetailsLinkParams ) =>
```
Comment (shendy-a8c) [inline client/components/details-link/index.tsx:23]: With rule @typescript-eslint/explicit-module-boundary-types on, linter complains this exported function is missing return type.

```suggestion
const DetailsLink = ( {
	id,
	parentSegment,
}: DetailsLinkParams ): JSX.Element | null =>
```

File: package.json [hunk 4]
```diff
@@ -102,6 +102,7 @@
     "@wordpress/i18n": "3.5.0",
     "@wordpress/jest-preset-default": "4.0.0",
     "@wordpress/scripts": "12.4.0",
+    "@wordpress/url": "2.22.0",
```
Comment (shendy-a8c) [inline package.json:105]: I see references to `@wordpress/url` but you just added it here. How did it work before?

**Review by shendy-a8c — COMMENTED** [review]
Looks good. I have some questions 👆.

Also, there is a failing test but I see it's already discussed over Slack.

**Comment by shendy-a8c** [comment]
There is subtle difference between develop and this branch when I open the deposit list page, but the one of this branch functions well.

develop:
Screen Shot 2021-03-30 at 17 55 34


This branch:
Screen Shot 2021-03-30 at 17 55 26

**Comment by reykjalin** [comment]
> There is subtle difference between develop and this branch when I open the deposit list page

@shendy-a8c They look the same for me 🤔 . You're talking about the first column being shaded, correct? I wonder if maybe there was some style change introduced with #1230 that maybe you haven't pulled to your local `develop` yet?

All the tables on my end have the first column shaded 🤷

**Comment by shendy-a8c** [comment]
Hmm.. could be. I thought I already pulled `develop`. Anyway, the subtle differences I see, which I don't think is a big deal, just want to bring it up:
- 1st column is shaded.
- There is an arrow indicating the list is sorted by that column.
- Header row is darker.
- 'Amount' header is not right-aligned the same way.

Probably something on my end. Will let you know when I find out.

**Comment by reykjalin** [comment]
Too old to keep working on, so closing this.

**Pull request closed without merging**